### PR TITLE
[DispatchCreation] Remove unit dim from attn mask

### PR DESCRIPTION
**Pull request by IanWood1**
Removes unit dims from the mask operand of the attention op. The goal is that this will allow the mask computation to be fused with the attention op.

### CODE REVIEW

File: compiler/src/iree/compiler/DispatchCreation/test/fold_unit_dims.mlir [hunk 1]
```diff
@@ -127,3 +127,66 @@ util.func public @scatter(%arg0 : tensor<4xi64>, %arg1 : tensor<4x1xi32>, %arg2
...
+  } -> tensor<8x4x1x128xf32>
+  util.return %1 : tensor<8x4x1x128xf32>
+}
+// CHECK-LABEL: func public @attention_mask_multi_m_dims
+// CHECK-SAME:    %[[ARG0:[a-zA-Z0-9]+]]
+// CHECK-SAME:    %[[ARG1:[a-zA-Z0-9]+]]
+// CHECK-SAME:    %[[ARG2:[a-zA-Z0-9]+]]
+// CHECK-SAME:    %[[ARG3:[a-zA-Z0-9]+]]
+// CHECK-SAME:    %[[ARG4:[a-zA-Z0-9]+]]
+//       CHECK:   %[[COLLAPSED:.+]] = tensor.collapse_shape %[[ARG4]]
```
Comment (MaheshRavishankar) [inline compiler/src/iree/compiler/DispatchCreation/test/fold_unit_dims.mlir:156]: You probably want to check you get the attention mask you want as well. Cause I think `d2` gets dropped in the final indexing map.

**Review by MaheshRavishankar — APPROVED** [review]
